ref(flags): support sort and provider filter on logs endpoint

aliu39 opened the pull request:
We need this for https://github.com/getsentry/sentry/issues/85219, to get the most recent X logs in reverse order.
Pagination test is added as an example of how to paginate - from the frontend this will mostly be handled by [pagination.tsx](https://github.com/getsentry/sentry/blob/02165b360d750a982126afedccb6c3faac9cf49c/static/app/components/pagination.tsx#L1-L108)

aliu39 committed:
ref(flags): support sort on FlagLogIndexEndpoint

aliu39 committed:
Rm comment

aliu39 committed:
Camel case update

aliu39 committed:
Update test name

cmanallen commented on src/sentry/flags/endpoints/logs.py:
```diff
@@ -67,6 +68,12 @@ def get(self, request: Request, organization: Organization) -> Response:
 
+        sort = request.GET.get("sort")
+        # Support camel case since it's used by our response serializer.
+        sort = camel_to_snake_case(sort) if isinstance(sort, str) else "created_at"
```
What happens if you sort by a column that doesn't exist?

aliu39 replied:
500

aliu39 replied:
I can try catching those errors and raising 400?

JoshFerge replied:
can you use a request serializer for the query parameters?

aliu39 committed:
Use request serializer and support provider filter

aliu39 commented:
I updated with a request serializer, although the column existence check/raise 400 is done outside of that.
Also added a provider filter if y'all don't mind me squeezing it in :)

aliu39 committed:
start end comment

aliu39 committed:
Revert secrets

aliu39 committed:
Support 'unknown' filter for provider=None

cmanallen commented on src/sentry/flags/endpoints/logs.py:
```diff
@@ -47,6 +57,26 @@ def serialize(self, obj, attrs, user, **kwargs) -> FlagAuditLogModelSerializerRe
+
+    # Support camel case since it's used by our response serializer.
+    def validate_sort(self, value: str | None) -> str | None:
+        return camel_to_snake_case(value) if value else None
```
This would be a great place to check if the sort value is in a pre-approved list of fields.

cmanallen reviewed (approved):
I'm not going to block this PR any longer but I think it would be a good thing to assert the sort value is one of a few approved fields rather than relying on a FieldError to catch bad input.  In other words, explicit is better than implicit.

aliu39 committed:
Validate sort explicitly

aliu39 committed:
fix mypy
